File: ios/VehicleTracker/Engine/ShapeGeometry.swift
import Foundation

/// A WGS84 coordinate in degrees.
nonisolated struct GeoPoint: Sendable, Equatable, Hashable {
    var lat: Double
    var lon: Double

    init(_ lat: Double, _ lon: Double) {
        self.lat = lat
        self.lon = lon
    }
}

/// Great-circle helpers: the same formulas and constants as the server's
/// `rider/shape.go`, so the phone's distances agree with the server's.
nonisolated enum Geo {
    static let earthRadiusM = 6_371_000.0
    /// Local equirectangular scale factor.
    static let metresPerDegree = 111_320.0

    static func distance(_ a: GeoPoint, _ b: GeoPoint) -> Double {
        let lat1 = rad(a.lat), lat2 = rad(b.lat)
        let dLat = lat2 - lat1
        let dLon = rad(b.lon - a.lon)
        let h = sin(dLat / 2) * sin(dLat / 2) + cos(lat1) * cos(lat2) * sin(dLon / 2) * sin(dLon / 2)
        return 2 * earthRadiusM * asin(min(1, sqrt(h)))
    }

    /// Initial bearing from a to b, degrees clockwise from north in [0, 360).
    static func initialBearing(_ a: GeoPoint, _ b: GeoPoint) -> Double {
        let lat1 = rad(a.lat), lat2 = rad(b.lat)
        let dLon = rad(b.lon - a.lon)
        let y = sin(dLon) * cos(lat2)
        let x = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(dLon)
        let deg = atan2(y, x) * 180 / .pi
        return (deg + 360).truncatingRemainder(dividingBy: 360)
    }

    static func rad(_ deg: Double) -> Double { deg * .pi / 180 }
}

/// The result of projecting a point onto a shape.
nonisolated struct Projection: Sendable, Equatable {
    /// Metres from the shape start.
    var alongShape: Double
    /// Metres from the point to the closest point on the shape.
    var distanceToShape: Double
}

/// A polyline with precomputed cumulative distances; a port of the server's
/// `rider.ShapeGeom`, including its hinted projection that keeps loops and
/// out-and-backs from snapping to the wrong pass.
nonisolated struct ShapeGeometry: Sendable {
    /// Minimum width, in metres, of the band of local minima a hint may choose
    /// between. It lets a hint choose between passes of a loop that share a
    /// point, where the closest distance is near zero.
    static let hintCandidateBand = 30.0
    /// How much further behind the hint a candidate pass must be, relative to
    /// one ahead of it, before it wins: a vehicle moves forward along its trip.
    static let backwardHintWeight = 2.0

    let points: [GeoPoint]
    /// cumulative[i] = metres from points[0] to points[i].
    let cumulative: [Double]
    let length: Double

    private let origin: GeoPoint
    private let cosLat: Double
    private let localX: [Double]
    private let localY: [Double]

    /// Nil for fewer than two points, which is no shape at all.
    init?(points: [GeoPoint]) {
        guard points.count >= 2 else { return nil }
        self.points = points
        origin = points[0]
        let meanLat = points.reduce(0) { $0 + $1.lat } / Double(points.count)
        cosLat = cos(Geo.rad(meanLat))

        var cumulative = [Double](repeating: 0, count: points.count)
        for i in 1..<points.count {
            cumulative[i] = cumulative[i - 1] + Geo.distance(points[i - 1], points[i])
        }
        self.cumulative = cumulative
        length = cumulative[points.count - 1]

        var xs: [Double] = []
        var ys: [Double] = []
        xs.reserveCapacity(points.count)
        ys.reserveCapacity(points.count)
        for p in points {
            xs.append((p.lon - points[0].lon) * cosLat * Geo.metresPerDegree)
            ys.append((p.lat - points[0].lat) * Geo.metresPerDegree)
        }
        localX = xs
        localY = ys
    }

    /// The position of p along the shape. Without a hint the globally closest
    /// segment wins. With a hint (the previous along-shape distance) the local
    /// minimum nearest the hint wins instead, among those within
    /// max(2×best + 1, hintCandidateBand) metres of the best, with backwards
    /// distance weighted by backwardHintWeight.
    func project(_ p: GeoPoint, hint: Double?) -> Projection {
        let segments = points.count - 1
        let (px, py) = local(p)

        guard let hint else {
            var bestDist = Double.infinity
            var bestAlong = 0.0
            for i in 0..<segments {
                let (d, along) = projectOnto(i, px, py)
                if d < bestDist {
                    bestDist = d
                    bestAlong = along
                }
            }
            return Projection(alongShape: bestAlong, distanceToShape: bestDist)
        }

        var dists = [Double](repeating: 0, count: segments)
        var alongs = [Double](repeating: 0, count: segments)
        var best = 0
        for i in 0..<segments {
            (dists[i], alongs[i]) = projectOnto(i, px, py)
            if dists[i] < dists[best] { best = i }
        }

        var chosen = best
        let threshold = max(2 * dists[best] + 1, Self.hintCandidateBand)
        var closest = Double.infinity
        for i in 0..<segments {
            if dists[i] > threshold || !Self.isLocalMin(dists, i) { continue }
            var delta = alongs[i] - hint
            if delta < 0 { delta = -delta * Self.backwardHintWeight }
            if delta < closest {
                closest = delta
                chosen = i
            }
        }
        return Projection(alongShape: alongs[chosen], distanceToShape: dists[chosen])
    }

    /// The coordinate `along` metres into the shape, clamped to [0, length].
    func point(at along: Double) -> GeoPoint {
        let (i, t) = segment(at: along)
        let a = points[i], b = points[i + 1]
        return GeoPoint(a.lat + t * (b.lat - a.lat), a.lon + t * (b.lon - a.lon))
    }

    /// The bearing of the segment containing `along`.
    func bearing(at along: Double) -> Double {
        let (i, _) = segment(at: along)
        return Geo.initialBearing(points[i], points[i + 1])
    }

    /// Distance from the local-plane point to segment i, and the along-shape
    /// distance of the closest point on it.
    private func projectOnto(_ i: Int, _ px: Double, _ py: Double) -> (Double, Double) {
        let ax = localX[i], ay = localY[i]
        let dx = localX[i + 1] - ax, dy = localY[i + 1] - ay
        var t = 0.0
        let lenSq = dx * dx + dy * dy
        if lenSq > 0 {
            t = min(1, max(0, ((px - ax) * dx + (py - ay) * dy) / lenSq))
        }
        let dist = hypot(px - (ax + t * dx), py - (ay + t * dy))
        return (dist, cumulative[i] + t * (cumulative[i + 1] - cumulative[i]))
    }

    /// The index of the segment containing `along` and the fraction into it,
    /// clamping `along` to [0, length].
    private func segment(at along: Double) -> (Int, Double) {
        let last = points.count - 2
        if along >= length { return (last, 1) }
        if along <= 0 { return (0, 0) }
        for i in 0...last where along < cumulative[i + 1] {
            let segLen = cumulative[i + 1] - cumulative[i]
            return segLen <= 0 ? (i, 0) : (i, (along - cumulative[i]) / segLen)
        }
        return (last, 1)
    }

    /// p in metres in the shape's equirectangular projection.
    private func local(_ p: GeoPoint) -> (Double, Double) {
        ((p.lon - origin.lon) * cosLat * Geo.metresPerDegree,
         (p.lat - origin.lat) * Geo.metresPerDegree)
    }

    /// Whether dists[i] is no greater than its neighbours.
    private static func isLocalMin(_ dists: [Double], _ i: Int) -> Bool {
        if i > 0 && dists[i] > dists[i - 1] { return false }
        if i < dists.count - 1 && dists[i] > dists[i + 1] { return false }
        return true
    }
}

File: android/app/src/main/kotlin/org/onebusaway/vehicletracker/engine/ShapeGeometry.kt
package org.onebusaway.vehicletracker.engine

import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min

/** The result of projecting a point onto a shape. */
data class Projection(
    /** Metres from the shape start. */
    val alongShape: Double,
    /** Metres from the point to the closest point on the shape. */
    val distanceToShape: Double,
)

/**
 * A polyline with precomputed cumulative distances: a port of the server's `rider.ShapeGeom`
 * (`rider/shape.go`), by way of `ShapeGeometry.swift`, including its hinted projection that keeps
 * loops and out-and-backs from snapping to the wrong pass.
 */
class ShapeGeometry private constructor(val points: List<GeoPoint>) {
    /** cumulative[i] = metres from points[0] to points[i]. */
    val cumulative: List<Double>
    val length: Double

    /** The local-projection origin, points[0]. */
    private val origin = points[0]

    /** The cosine of the mean latitude. */
    private val cosLat = cos(Geo.rad(points.sumOf { it.lat } / points.size))

    // localX and localY are the points projected onto the local plane once, at construction:
    // project runs on every fix, and re-deriving the vertices from degrees on each call was the
    // bulk of its work.
    private val localX = DoubleArray(points.size)
    private val localY = DoubleArray(points.size)

    init {
        val cumulative = DoubleArray(points.size)
        for (i in 1 until points.size) {
            cumulative[i] = cumulative[i - 1] + Geo.distance(points[i - 1], points[i])
        }
        this.cumulative = cumulative.asList()
        length = cumulative[points.size - 1]

        for ((i, p) in points.withIndex()) {
            val (x, y) = local(p)
            localX[i] = x
            localY[i] = y
        }
    }

    /**
     * The position of [p] along the shape. Distances are measured in a local equirectangular
     * projection. When [hint] is null the globally closest segment wins, found in one pass that
     * keeps nothing. With a hint — the path every fix after a trip's first match takes — the local
     * minimum nearest the hinted distance along the shape wins instead, backwards distance counting
     * for more than forwards, which keeps loops and out-and-backs from snapping to the wrong pass;
     * that needs every segment's distance at once, so it keeps the two arrays.
     */
    fun project(p: GeoPoint, hint: Double?): Projection {
        val segments = points.size - 1
        val (px, py) = local(p)

        if (hint == null) {
            var best = Projection(alongShape = 0.0, distanceToShape = Double.POSITIVE_INFINITY)
            for (i in 0 until segments) {
                val onto = projectOnto(i, px, py)
                if (onto.distanceToShape < best.distanceToShape) best = onto
            }
            return best
        }

        val dists = DoubleArray(segments)
        val alongs = DoubleArray(segments)
        var best = 0
        for (i in 0 until segments) {
            val onto = projectOnto(i, px, py)
            dists[i] = onto.distanceToShape
            alongs[i] = onto.alongShape
            if (dists[i] < dists[best]) best = i
        }

        var chosen = best
        val threshold = max(2 * dists[best] + 1, HINT_CANDIDATE_BAND)
        var closest = Double.POSITIVE_INFINITY
        for (i in 0 until segments) {
            if (dists[i] > threshold || !isLocalMin(dists, i)) continue
            var delta = alongs[i] - hint
            if (delta < 0) delta = -delta * BACKWARD_HINT_WEIGHT
            if (delta < closest) {
                closest = delta
                chosen = i
            }
        }
        return Projection(alongShape = alongs[chosen], distanceToShape = dists[chosen])
    }

    /** The coordinate [along] metres into the shape, clamped to [0, length]. */
    fun pointAt(along: Double): GeoPoint {
        val (i, t) = segmentAt(along)
        val a = points[i]
        val b = points[i + 1]
        return GeoPoint(a.lat + t * (b.lat - a.lat), a.lon + t * (b.lon - a.lon))
    }

    /** The bearing of the segment containing [along]. */
    fun bearingAt(along: Double): Double {
        val (i, _) = segmentAt(along)
        return Geo.initialBearing(points[i], points[i + 1])
    }

    /**
     * The projection of the local-plane point ([px], [py]) onto segment [i]: its distance from the
     * segment, and the along-shape distance of the closest point on it.
     */
    private fun projectOnto(i: Int, px: Double, py: Double): Projection {
        val ax = localX[i]
        val ay = localY[i]
        val dx = localX[i + 1] - ax
        val dy = localY[i + 1] - ay

        var t = 0.0
        val lenSq = dx * dx + dy * dy
        if (lenSq > 0) t = min(1.0, max(0.0, ((px - ax) * dx + (py - ay) * dy) / lenSq))
        return Projection(
            alongShape = cumulative[i] + t * (cumulative[i + 1] - cumulative[i]),
            distanceToShape = hypot(px - (ax + t * dx), py - (ay + t * dy)),
        )
    }

    /**
     * The index of the segment containing [along] and the fraction into that segment, clamping
     * [along] to [0, length].
     */
    private fun segmentAt(along: Double): Pair<Int, Double> {
        val last = points.size - 2
        if (along >= length) return last to 1.0
        if (along <= 0) return 0 to 0.0
        for (i in 0..last) {
            if (along < cumulative[i + 1]) {
                val segLen = cumulative[i + 1] - cumulative[i]
                if (segLen <= 0) return i to 0.0
                return i to (along - cumulative[i]) / segLen
            }
        }
        return last to 1.0
    }

    /** [p] in metres in the shape's equirectangular projection. */
    private fun local(p: GeoPoint): Pair<Double, Double> = Pair(
        (p.lon - origin.lon) * cosLat * Geo.METRES_PER_DEGREE,
        (p.lat - origin.lat) * Geo.METRES_PER_DEGREE,
    )

    companion object {
        /**
         * The minimum width, in metres, of the band of local minima a hint may choose between. It
         * lets a hint choose between passes of a loop that share a point, where the closest
         * distance is near zero and a purely proportional band would admit only the one pass.
         */
        private const val HINT_CANDIDATE_BAND = 30.0

        /**
         * How much further behind the hint a candidate pass must be, relative to one ahead of it,
         * before it wins. A vehicle on its trip moves forward along the shape: just past the
         * turnaround of an out-and-back the two legs are equally near the last match, and the one
         * ahead is the right one. A short step backwards on the same pass — GPS jitter, a bus
         * reversing into a bay — still beats a pass far ahead.
         */
        private const val BACKWARD_HINT_WEIGHT = 2.0

        /** Null for fewer than two points, which is no shape at all; the Go constructor panics. */
        fun of(points: List<GeoPoint>): ShapeGeometry? =
            if (points.size < 2) null else ShapeGeometry(points.toList())
    }
}

/** Whether dists[i] is no greater than its neighbours. */
private fun isLocalMin(dists: DoubleArray, i: Int): Boolean {
    if (i > 0 && dists[i] > dists[i - 1]) return false
    if (i < dists.size - 1 && dists[i] > dists[i + 1]) return false
    return true
}
